File: flink-app/src/main/java/com/example/ctr/domain/model/Event.java
package com.example.ctr.domain.model;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.util.Set;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@JsonIgnoreProperties(ignoreUnknown = true)
public class Event {
    @JsonProperty("event_id")
    private String eventId;

    @JsonProperty("event_type")
    private String eventType;

    @JsonProperty("product_id")
    private String productId;

    @JsonProperty("user_id")
    private String userId;

    @JsonProperty("timestamp")
    @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
    private LocalDateTime timestamp;

    @JsonProperty("session_id")
    private String sessionId;

    public boolean hasProductId() {
        return productId != null && !productId.isEmpty();
    }

    public boolean isValid() {
        return hasProductId()
                && timestamp != null
                && VALID_EVENT_TYPES.contains(eventType);
    }

    public long eventTimeMillisUtc() {
        if (timestamp == null) {
            return 0L;
        }
        return timestamp.atOffset(ZoneOffset.UTC).toInstant().toEpochMilli();
    }

    private static final Set<String> VALID_EVENT_TYPES = Set.of("view", "impression", "click");
}

File: flink-app/src/main/kotlin/com/example/ctr/domain/model/Event.kt
package com.example.ctr.domain.model

import com.fasterxml.jackson.annotation.JsonFormat
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonProperty
import java.time.LocalDateTime
import java.time.ZoneOffset

@JsonIgnoreProperties(ignoreUnknown = true)
data class Event(
    @JsonProperty("event_id")
    val eventId: String? = null,

    @JsonProperty("event_type")
    val eventType: String? = null,

    @JsonProperty("product_id")
    val productId: String? = null,

    @JsonProperty("user_id")
    val userId: String? = null,

    @JsonProperty("timestamp")
    @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
    val timestamp: LocalDateTime? = null,

    @JsonProperty("session_id")
    val sessionId: String? = null
) {

    fun hasProductId(): Boolean = !productId.isNullOrBlank()

    fun isValid(): Boolean = hasProductId() && timestamp != null && VALID_EVENT_TYPES.contains(eventType)

    fun eventTimeMillisUtc(): Long = timestamp?.atOffset(ZoneOffset.UTC)?.toInstant()?.toEpochMilli() ?: 0L

    companion object {
        private val VALID_EVENT_TYPES: Set<String> = setOf("view", "impression", "click")
    }
}
